Dockerfile

```diff
@@ -0,0 +1,57 @@
+# This dockerfile is used to test containerd within a container
+#
+# usage: 
+# 1.) docker build -t containerd-test -f Dockerfile .
+# 2.) docker run -it --privileged containerd-test
+# 3.) make test
+#
+
+FROM golang:1.9.2
+
+RUN apt-get update && apt-get install -y \
+	autoconf \
+	automake \
+	btrfs-tools \
+	curl \
+	g++ \
+	git \
+	libapparmor-dev \
+	libseccomp-dev \
+	libtool \
+	tar \
+	unzip
+
+# Install proto3
+RUN set -x \
+	&& mkdir -p /go/src/github.com/google \
+	&& cd /go/src/github.com/google \
+	&& git clone "https://github.com/google/protobuf.git" \
+	&& cd protobuf \
+	&& ./autogen.sh \
+	&& ./configure \
+	&& make \
+	&& make check \
+	&& make install \
+	&& ldconfig
+
+# Install runc
+RUN set -x \
+	&& mkdir -p /go/src/github.com/opencontainers \
+	&& cd /go/src/github.com/opencontainers \
+	&& git clone "https://github.com/opencontainers/runc.git" \
+	&& cd runc \
+	&& make runc install \
+	&& cp runc /go/bin
+
+# Install containerd
+RUN set -x \
+	&& mkdir -p /go/src/github.com/containerd \
+	&& cd /go/src/github.com/containerd \
+	&& git clone "https://github.com/containerd/containerd.git" \
+	&& cd containerd \
+	&& make binaries install
```

Comment:
I generally avoid this "git checkout of the source repo" in a Dockerfile. I think it's safe to assume that this Dockerfile will only be used with a build context of a `containerd` checkout, so it should use `COPY . /go/src/github.com/containerd/containerd` or use a bind mount to get a source.

`make binaries install` can be done as part of the `ENTRYPOINT` or omitted entirely.